File: advent-of-code/2020/Day10.kt
import java.lang.Integer.max

const val MAX_JUMP = 3

fun main() {
    var chain = mapLines { it.toInt() }.sorted().let { listOf(0) + it + listOf(it.last() + 3) }

    val first = chain.asSequence()
        .zipWithNext()
        .map { (a, b) -> b - a }
        .groupBy { it }
        .let { it[1]!!.size * it[3]!!.size }

    val ways = MutableList(chain.size) { 0L }
    ways[0] = 1L
    chain.asSequence().withIndex().drop(1).forEach { (index, el) ->
        ways[index] = (max(index - MAX_JUMP, 0)..index - 1).asSequence()
            .filter { el - chain[it] <= MAX_JUMP }
            .map { ways[it] }
            .sum()
    }
    val second = ways.last()

    println("$first $second")
}

File: advent-of-code/2020/Day10.ml
open Printf

let max_jump = 3;;

let rev_lines =
  let lines = ref [] in
  try
    while true; do
      lines := input_line stdin :: !lines
    done;
    !lines
  with End_of_file -> !lines;;

let adapters = List.map int_of_string rev_lines |> List.sort compare;;
let chain = 0 :: adapters @ [3 + List.hd (List.rev adapters)];;


let first =
  let left = List.rev chain |> List.tl |> List.rev in
  let right = List.tl chain in
  let diffs = List.map2 (-) right left in
  let find n = List.filter (fun a -> a == n) diffs |> List.length in
  find 1 * find 3;;

let ways = ref [1];;
let second =
  let rec sum = function
    | [] -> 0
    | h::t -> h + (sum t)
  in
  let rec index_list index acc =
    if acc == max_jump then []
    else
    let tail_rec = index_list (index + 1) (acc + 1) in
    if index < 0 then tail_rec
    else index :: tail_rec
  in
  let chain_iter index el =
    ways := !ways @ [
        index_list (index - max_jump + 1) 0
        |> List.filter (fun it -> el - List.nth chain it <= max_jump)
        |> List.map (List.nth !ways)
        |> sum
    ]
  in
  List.tl chain |> List.iteri chain_iter;
  List.rev !ways |> List.hd;;

Printf.printf "%d %d\n" first second;;
